#  Midterm TEST 1A 2024 
## Name <font color='red'> Santino Sini </font>
## Date <font color='red'> 3/6/2024 </font>
## Course<font color='red'> CSC305 CSC405 </font>
## Run the given code when requested, provide code when requested in the code cells, and answer questions inserting responses in appropriate place (red font or ?)



# Available Information about Data:
* Process trajectory data recorded as GPS trace
* Real data sets recorded by people carrying GPS enabled devices
* Data available over multiple years for people traveling in different modes like walking, riding bus etc.
* Most important part of data is latitude, longitude sequence with timestamp for each pair (forms trajectory data)
* Several interesting problems in Activity Monitoring are naturally related to such data sets  

* A smaller data set named "MysteriousData2" is provided along with this notebook



# Task 1 - Initial Code: Importing Libraries

<font color='blue'>
    
#### A) General Information: 
* Very often we need to import  additional programs that are organized into various libraries.

#### C) Run the given code: </font>

In [1]:
import pandas as pd  #imports the "pandas" library



#### E) Concept Understanding:

1.2 Why do we need specifically pandas library?

<font color='red'> The pandas library is used for creating and manipulating data frames from data. </font>
_________________________________


1.2 Why do we need specifically os library?

<font color='red'> os interacts with the operating system and creating files and directories. </font>
_________________________________


In [2]:
import os

In [3]:
# Write the code to import "numpy" library 
import numpy as np

# Task 2:  Identifying a File in a Hierarchical Directory

* Investigate "MysteriousData" directory and draw this directory hierarchy on a piece of paper or electronic document.
* Choose the first file from two csv files for user 121
* Store file name as string 
* Construct the full path to this file in the directory hierarchy as a string 
* Specify file identification as a string by concatenating the path with file name
* print file identification to check if it is correct.

#### C) Create and run the apprpriate  code: </font>

In [5]:
# How to reference a file - by a path concatenated with file name - we call it File Identification
#?
fileName = '200910051028v2.csv'
path = './MysteriousData2/121/'
FileIdentification1 = os.path.join(path, fileName)
print(FileIdentification1)


./MysteriousData2/121/200910051028v2.csv


#### D) Output Understanding:

2.1. Was the file identificator for the first file  according to your expectation?

<font color='red'> ./MysteriousData2/121/200910051028v2.csv
</font>

#### E) Concept Understanding:

2.2. Using the right click, invoke the file properties for this file in the Windows Explorer directory and file manager. 
* Select the security tab, and copy the object name and paste below.

<font color='red'>/home/santino/Documents/Current_Projects/classes/data_analytics/jupyter_notebooks/MysteriousData2/121</font>
_________________________________

2.3. Is the object name different than our file identifier? Why?  

<font color='red'> Yes, because './' is used to begin the path in the current working directory. </font>



#### G) Apply in typical situations:

2.4. Write the code to create the second file identification variable 
* call this variable FileIdentification2
* print the result

In [6]:
# How to reference a file - by a path concatenated with file name that we call it File Identification
fileName2 = '200910111153v2.csv'
path = './MysteriousData2/121/'
FileIdentification2 = os.path.join(path, fileName2)
print(FileIdentification2)


./MysteriousData2/121/200910111153v2.csv


#### E) Concept Understanding:

2.5. Using the right click, invoke the file properties for this file in the Windows Explorer directory and file manager. 
* Select the security tab, and copy the Object name and paste below.

<font color='red'> /home/santino/Documents/Current_Projects/classes/data_analytics/jupyter_notebooks/MysteriousData2/121 </font>
_________________________________



# Task 3: Reading csv files into Tables (Pandas Dataframes) with Fixed Path


#### C) Create  and Run the apprpriate code i.e. Write the code to load the first file of user 121 into a new dataframe called Points_User121_File2 and display it

In [7]:
# Load the first data file for the specific user and display a few rows 
Points_MysteriousUser120_File1 = pd.read_csv(FileIdentification1)
Points_MysteriousUser120_File1

Order   Latitude   Longitude  Unused     Date      Time
0        0  39.180741  117.143490       0  10/5/09  10:28:28
1        1  39.180634  117.143428       0  10/5/09  10:28:33
2        2  39.180551  117.143343       0  10/5/09  10:28:38
3        3  39.180478  117.143261       0  10/5/09  10:28:43
4        4  39.180337  117.143144       0  10/5/09  10:28:48
..     ...        ...         ...     ...      ...       ...
564    564  39.181190  117.143998       0  10/5/09  11:58:42
565    565  39.181147  117.144037       0  10/5/09  11:58:47
566    566  39.181153  117.144059       0  10/5/09  11:58:52
567    567  39.181115  117.144096       0  10/5/09  11:58:57
568    568  39.181031  117.144189       0  10/5/09  11:59:02

[569 rows x 6 columns]

#### D) Output Understanding:

3.1. What are column names?

<font color='red'> Order, Latitude, Longitude, Unused, Date, Time</font>

3.2. How many rows are reported?

<font color='red'>  567 </font>
_________________________________

3.3 Write the code to load the second file of user 121 into a new dataframe called Points_User121_File2 and display it


In [8]:
# Load data for a specific user and display some rows 
Points_MysteriousUser120_File2 = pd.read_csv(FileIdentification2)
Points_MysteriousUser120_File2

Order   Latitude   Longitude  Unused      Date      Time
0         0  39.980956  116.329595       0  10/11/09  11:53:42
1         1  39.980913  116.329603       0  10/11/09  11:53:47
2         2  39.980920  116.329612       0  10/11/09  11:53:52
3         3  39.980929  116.329621       0  10/11/09  11:53:57
4         4  39.980929  116.329621       0  10/11/09  11:54:00
...     ...        ...         ...     ...       ...       ...
2451   2451  39.980755  116.329306       0  10/11/09  21:57:57
2452   2452  39.980814  116.329182       0  10/11/09  21:58:02
2453   2453  39.980723  116.329096       0  10/11/09  21:58:07
2454   2454  39.980412  116.329317       0  10/11/09  21:58:12
2455   2455  39.980230  116.329482       0  10/11/09  21:58:17

[2456 rows x 6 columns]

<font color='blue'>

### We can define a functions  that will read data from a file by hard coding the path if the directory structure is fixed for our application.

#### For example we can define a function  that will read data from a file by specifying value for the name of the file

####:
* Complete and Execute the function readDataFromFile121()   </font>

In [13]:
#Loads data from a file from a specific user in the directory  
def readDataFromFile121(fileName):
    path = './MysteriousData2/121/'
    FileIdentification = os.path.join(path, fileName)
    dataframe = pd.read_csv(FileIdentification)  
    return(dataframe)


In [20]:
#?
fileName1 = '200910051028v2.csv'
mysteryDataframe1 = readDataFromFile121(fileName1)
mysteryDataframe1

Order   Latitude   Longitude  Unused     Date      Time
0        0  39.180741  117.143490       0  10/5/09  10:28:28
1        1  39.180634  117.143428       0  10/5/09  10:28:33
2        2  39.180551  117.143343       0  10/5/09  10:28:38
3        3  39.180478  117.143261       0  10/5/09  10:28:43
4        4  39.180337  117.143144       0  10/5/09  10:28:48
..     ...        ...         ...     ...      ...       ...
564    564  39.181190  117.143998       0  10/5/09  11:58:42
565    565  39.181147  117.144037       0  10/5/09  11:58:47
566    566  39.181153  117.144059       0  10/5/09  11:58:52
567    567  39.181115  117.144096       0  10/5/09  11:58:57
568    568  39.181031  117.144189       0  10/5/09  11:59:02

[569 rows x 6 columns]

#### D) Apply in typical situations:
3.6. Write the code to load the second file of user 121 into a new dataframe called Points_User121_File2 using the function readDataFromFile121()

In [21]:
#?
fileName2 = '200910111153v2.csv'
mysteryDataframe2 = readDataFromFile121(fileName2)
mysteryDataframe2

Order   Latitude   Longitude  Unused      Date      Time
0         0  39.980956  116.329595       0  10/11/09  11:53:42
1         1  39.980913  116.329603       0  10/11/09  11:53:47
2         2  39.980920  116.329612       0  10/11/09  11:53:52
3         3  39.980929  116.329621       0  10/11/09  11:53:57
4         4  39.980929  116.329621       0  10/11/09  11:54:00
...     ...        ...         ...     ...       ...       ...
2451   2451  39.980755  116.329306       0  10/11/09  21:57:57
2452   2452  39.980814  116.329182       0  10/11/09  21:58:02
2453   2453  39.980723  116.329096       0  10/11/09  21:58:07
2454   2454  39.980412  116.329317       0  10/11/09  21:58:12
2455   2455  39.980230  116.329482       0  10/11/09  21:58:17

[2456 rows x 6 columns]

# Task 4: Saving Tables (Pandas Dataframes) into csv files 

<font color='blue'>


Create and run the code to upload  two discussed dataframes into a csv text file in the current directory.



In [22]:
#?
mysteryDataframe1.to_csv('mysteryDataframe1', index=False) 
mysteryDataframe2.to_csv('mysteryDataframe2', index=False) 

#### F) Output Understanding:

4.1. Using the right click, invoke the file properties for the first newly created file in the Windows Explorer directory. 
* Select the security tab, and copy the object name and paste below.

<font color='red'> /home/santino/Documents/Current_Projects/classes/data_analytics/jupyter_notebooks/mysteryDataframe1 </font> 


4.2. Using the right click, invoke the file properties for the first newly created file in the Windows Explorer directory. 
* Select the security tab, and copy the object name and paste below.

<font color='red'> /home/santino/Documents/Current_Projects/classes/data_analytics/jupyter_notebooks/mysteryDataframe2 </font> 

